tests/collector/integration.js

```diff
@@ -212,1238 +276,1559 @@ const nockConfig = [
+          .then(() => u.awaitHeartbeat())
+          .then(() => u.postGenerator(gen4))
 
-      it('pause - heartbeats continue, collection stops', () =>
-        Promise.resolve()
-
-        .then(() => u.doStart(coll1))
-        .then(() => Promise.join(
-          u.awaitHeartbeat(),
-          u.awaitBulkUpsert(),
-        ))
-
-        .then(() => u.postStatus('pause', coll1))
-        .then(() => u.expectHeartbeatStatus(coll1, 'Paused'))
-        .then(() =>
-          u.awaitBulkUpsert()
-            .should.eventually.be.rejectedWith(Promise.TimeoutError),
-        )
-
-        .then(() => u.postStatus('resume', coll1))
-        .then(() => u.expectHeartbeatStatus(coll1, 'Running'))
-        .then(() =>
-          u.awaitBulkUpsert(),
-        )
-      );
-    });
-  });
+          .then(() => u.awaitBulkUpsert())
+          .then(({ req }) => {
+            expect(req.body).to.deep.equal([
+              {
+                name: 'sub1_1|asp1',
+                messageCode: 'ERROR',
+                messageBody: 'Error getting OAuth token (method=ownerPassword ' +
+                  'url=http://www.example.com///login): Unauthorized',
+                value: 'ERROR',
+              },
+              {
+                name: 'sub1_2|asp1',
+                messageCode: 'ERROR',
+                messageBody: 'Error getting OAuth token (method=ownerPassword ' +
+                  'url=http://www.example.com///login): Unauthorized',
+                value: 'ERROR',
+              },
+              {
+                name: 'sub1_3|asp1',
+                messageCode: 'ERROR',
+                messageBody: 'Error getting OAuth token (method=ownerPassword ' +
+                  'url=http://www.example.com///login): Unauthorized',
+                value: 'ERROR',
+              },
+            ]);
+          });
+        });
 
-  describe('assignment >', () => {
-    describe('generator updated >', () => {
-      it('generators created, assigned to collectors', () =>
-        Promise.join(
-          u.doStart(coll1)
-            .then(() => u.patchCollectorGroup(cg1.name, { collectors: [coll1] }))
-            .then(() => u.awaitHeartbeat(coll1)),
-          u.doStart(coll2)
-            .then(() => u.patchCollectorGroup(cg2.name, { collectors: [coll2] }))
-            .then(() => u.awaitHeartbeat(coll2)),
-        )
-
-        .then(() => Promise.join(
-          u.awaitHeartbeat(coll1)
-            .then(() => u.postGenerator(gen1))
-            .then(() => u.expectHeartbeatGens(coll1, { added: [gen1] })),
-          u.awaitHeartbeat(coll2)
-            .then(() => u.postGenerator(gen2))
-            .then(() => u.expectHeartbeatGens(coll2, { added: [gen2] })),
-        ))
-
-        .then(() => Promise.join(
-          u.expectBulkUpsertSamples(coll1, '60s', ['sub1|asp1']),
-          u.expectBulkUpsertSamples(coll2, '60s', ['sub2|asp2']),
-        ))
-      );
+        it('only gets token once, uses it for all subsequent requests', () => {
```

Comment:
@pallavi2209 this test verifies that the collector doesn't send multiple login requests.
I confirmed that this fails with the old version of the collector (before your fix), with a login count of 6.

Reply:
wow - these 3 tests! Thank you for the effort @jgraff2